### PR TITLE
Auth Implementation  /  Implementación De Autenticación

### PR DESCRIPTION
**Pull request by Israel-Laguan**
### Description:

Our app will require later to connect users and to hide certain information based on roles. This is known as authentication and authorization or in short Auth.

In NextJs we have a time-saving library called [NextAuth](https://next-auth.js.org/), this PR is to leverage this library and explore how its use of it can be on our app. 

#### Type of change:
- [x] New feature (non-breaking change which adds functionality).
- [x] This change requires a documentation update.

### How to test it:
If applicable, describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.

1. Go to ['/ingresar'](https://deploy-preview-33--did-admin-panel.netlify.app/ingresar)
2. Add a new user in [the table `users` in Hasura](https://did-db.herokuapp.com/console/data/did-db/schema/public/tables/users/insert), with both `username` and `password` as the email, then enter those in the `/ingresar` view
3. Click on 'Ingresar'.
4. Should get one message about a user found on DB, and then a prompt to check email.
5. Inspect your email to verify if you have a magic link
image
6. You should be able to be logged in to the app if you verify the link from the mail.
7. notice your user info is in session "Storage".

### Checklist:
- [x] Doesn't break the current code.
- [x] Passes linters and test, also is building.
- [x] Doesn't have spelling or grammatical problems.
- [x] Doesn't have unnecessary comments or debugging code.
- [x] The branch is updated with the last dev branch changes.

----

# Spanish Version:

### Descripción:

Nuestra aplicación requerirá más tarde conectar a los usuarios y ocultar cierta información según los roles. Esto se conoce como autenticación y autorización o, abreviado, Auth.

En NextJs tenemos una biblioteca que ahorra tiempo llamada [NextAuth](https://next-auth.js.org/). Este PR es para aprovechar esta biblioteca y explorar cómo se puede usar en nuestra aplicación.

#### Tipo de cambio:
- [x] Nueva función (cambio continuo que agrega funcionalidad).
- [x] Este cambio requiere una actualización de la documentación.

### Cómo probarlo:
Si corresponde, describa las pruebas que ejecutó para verificar sus cambios. Proporcione instrucciones para que podamos reproducir.

1. Vaya a ['/ingresar'](https://deploy-preview-33--did-admin-panel.netlify.app/ingresar)
2. Agregue un nuevo usuario en [la tabla `users` en Hasura](https://did-db.herokuapp.com/console/data/did-db/schema/public/tables/users/insert), con ambos `username` y `password` como correo electrónico. Luego, ingréselos en la vista `/ingresar`, 
3. Haga clic en 'Ingresar'.
4. Debería recibir un mensaje sobre un usuario encontrado en la base de datos y luego un aviso para revisar el correo electrónico.
5. Inspeccione su correo electrónico para verificar si tiene un enlace mágico.
imagen
6. Debería poder iniciar sesión en la aplicación si verifica en el enlace del correo.
7. Observe que su información de usuario está en sesión "Storage".

### Lista de Verificación:
- [x] No rompe el código actual.
- [x] Pasa linters y prueba. 
- [x] No tiene problemas ortográficos ni gramaticales.
- [x] No tiene comentarios innecesarios ni código de depuración.
- [x] La rama se actualiza con los últimos cambios de rama dev.

### CODE REVIEW

File: pages/api/auth/[...nextauth].js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import NextAuth from 'next-auth'
+import EmailProvider from 'next-auth/providers/email'
+
+export default NextAuth({
+  // Configure one or more authentication providers
+  providers: [
+    EmailProvider({
+      server: {
+        host: process.env.EMAIL_SERVER_HOST,
+        port: +process.env.EMAIL_SERVER_PORT,
```
Comment (Katsu08) [inline pages/api/auth/[...nextauth].js:10]: This `+` is a typo?

File: tsconfig.json [hunk 2]
```diff
@@ -8,7 +9,7 @@
     ],
     "allowJs": true,
     "skipLibCheck": true,
-    "strict": true,
+    "strict": false,
```
Comment (Katsu08) [inline tsconfig.json:12]: Is this change intentional?

**Review by Katsu08 — APPROVED** [review]
Just small doubts.

This PR can/need be tested? How could I do it?

**Review by e-azocar — COMMENTED** [review]
I've received this error on click sign in button
image

File: pages/api/auth/[...nextauth].js (hunk 1 above)
Comment (Israel-Laguan) [reply pages/api/auth/[...nextauth].js:10]: Ah no, is casting to a Number, if you think is confusing I can do a more explicit casting

**Comment by Israel-Laguan** [comment]
> I've received this error on click sign in button

So the email implementation is too much involved and means a lot of changes to make, so I decided to ditch that solution and use an alternative: current user verification plus a magic link to the email.

Also, people can be whitelisted to avoid extra people entering as admin. With magic, we can use a lot of fun stuff like several auth methods or modify the look and feel of the pop-up.

Please tell me how do you feel about the experience and how it can be improved.

File: tsconfig.json (hunk 2 above)
Comment (Israel-Laguan) [reply tsconfig.json]: I was checking and this change is not needed, so I reverted

**Comment by Israel-Laguan** [comment]
@victor-heliomar @e-azocar I think you 2 have to create credentials to complete the flow. We can talk about that in a meet.

File: views/Login/index.tsx [hunk 3]
```diff
@@ -1,52 +1,39 @@
...
-        variables: {
-          error: 'Unexpected state',
-          origin: 'ADMIN',
-          type: 'UNEXPECTED',
-          codeLocation: 'views::Login::L41',
-        },
-      })
+      logError(Error('Unexpected state'), 'views::Login::L32', 'UNEXPECTED')
     }
+    toast.success('Usuario encontrado en base de datos, procediendo a verificar sesión...', {
```
Comment (Israel-Laguan) [inline views/Login/index.tsx:34]: Move to correct verification

File: hooks/useError.tsx [hunk 4]
```diff
@@ -4,7 +4,26 @@ import { INSERT_ERROR } from '../services/GraphQL/mutations/errors'
...
+          error: JSON.stringify(error),
+          origin: 'ADMIN',
+          type,
+          codeLocation,
+        },
+      })
+    }
+  }
+
+  return [logError]
```
Comment (Israel-Laguan) [inline hooks/useError.tsx:26]: Was thinking on add more methods related to log, debug and show notifications/toasts. Can you guys help me to think about it?

Comment (e-azocar) [reply hooks/useError.tsx:26]: I think `logError` could return a `show` property, and if this is `true` shows a toast

Comment (Israel-Laguan) [reply hooks/useError.tsx:26]: What about something like "showNotification" or similar?

Comment (Israel-Laguan) [reply hooks/useError.tsx]: I updated the methods, please check again

File: views/Login/Layout.tsx [hunk 5]
```diff
@@ -6,7 +6,8 @@ import styles from './Login.module.scss'
 import LoginForm from './LoginForm'
 
 interface LoginFormProps {
-  onSubmit: (formData: LoginInput) => void
+  // eslint-disable-next-line no-unused-vars
+  onSubmit: (_formData: LoginInput) => void
```
[the remark this answers is not in the document]
Comment (Israel-Laguan) [reply views/Login/Layout.tsx:10]: Yes, it have this error 😢 
image

I tried to avoid this error in eslint conf and updated so is not reporting.

File: services/GraphQL/queries/users.ts [hunk 6]
```diff
@@ -69,14 +68,19 @@ export const GET_MEMBERS = gql`
         type
         updated_at
       }
+      user {
+        avatar_url
+        is_active
+        type
+        username
+      }
```
[the remark this answers is not in the document]
Comment (Israel-Laguan) [reply services/GraphQL/queries/users.ts]: I tried, I have some ideas to improve fragments but maybe in another PR

**Comment by Israel-Laguan** [comment]
I added test users in Hasura, we can check on that on a meet

**Comment by victor-heliomar** [comment]
@Israel-Laguan I think it's good to see him at a meet. 

Today in the afternoon I try to run the code to test the PR

File: views/Login/index.tsx [hunk 7]
```diff
@@ -1,51 +1,49 @@
...
+    } else {
+      if (data?.users?.length > 1) {
+        warn(
+          'Login:useEffect',
+          'Se encontraron varios usuarios en base de datos, usando el primer email. Informe al servicio técnico del incidente...',
+          'INPUT',
+        )
+      }
+      log('Login:useEffect', 'Usuario encontrado en base de datos, procediendo a verificar sesión...', 'SUCCESS')
+      if (data.users[0]?.is_active && data.users[0]?.member_information?.email) {
```
Comment (victor-heliomar) [inline views/Login/index.tsx:32]: This case can occurs if a user have the same member code and the same password

File: views/Login/index.tsx [hunk 8]
```diff
@@ -1,51 +1,49 @@
...
+        'INPUT',
+      )
+    } else {
+      if (data?.users?.length > 1) {
+        warn(
+          'Login:useEffect',
+          'Se encontraron varios usuarios en base de datos, usando el primer email. Informe al servicio técnico del incidente...',
+          'INPUT',
+        )
+      }
```
Comment (victor-heliomar) [inline views/Login/index.tsx:30]: This case can occurs if a user have the same member code and the same password

Comment (victor-heliomar) [reply views/Login/index.tsx:30]: But shouldn't be possible since the database should have the username as a unique primary key field

File: views/Login/index.tsx [hunk 9]
```diff
@@ -1,51 +1,49 @@
...
+      log('Login:useEffect', 'Usuario encontrado en base de datos, procediendo a verificar sesión...', 'SUCCESS')
+      if (data.users[0]?.is_active && data.users[0]?.member_information?.email) {
+        magicLink(data.users[0])
+      } else {
+        warn('Login:useEffect', 'Usuario no activo, contactar con el administrador...', 'AUTHORIZATION')
+      }
```
Comment (victor-heliomar) [inline views/Login/index.tsx:36]: I think you've missed request is_active from the database 

image

File: services/GraphQL/queries/users.ts [hunk 10]
```diff
@@ -1,73 +1,63 @@
...
+          is_active: { _eq: true }
+        }
+      }
+    ) {
+      ${usersInfo}
+      member_information {
+        ${memberInfo}
+      }
     }
   }
```
Comment (victor-heliomar) [inline services/GraphQL/queries/users.ts:23]: You're already filtering the users to get just the active users, I think is not needed to check here: 

https://github.com/Azordev/did-admin-panel/blob/b0660e259cbc9c9b4afa8bc028255306de16619e/views/Login/index.tsx#L32-L36

File: components/LoginButton.tsx [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+      </>
+    )
+  }
+  return (
+    <>
+      Not signed in <br />
+      <button onClick={() => signIn()}>Sign in</button>
+    </>
+  )
+}
```
Comment (victor-heliomar) [inline components/LoginButton.tsx:19]: nit: 

Just a proposal: 

```suggestion
   const onClick = session ? signOut : signIn;
   
   const captionMessage = session ? `Signed in as ${session.user?.email}` : "Not signed in";
   
   const buttonMessage = session ? "Sign out" : "Sign in";

    return (
      <>
        {captionMessage} <br />
        <button onClick={onClick}>{buttonMessage}</button>
      </>
    )
}
```

File: pages/index.tsx [hunk 12]
```diff
@@ -1,9 +1,21 @@
...
+  return (
+    <div>
+      <h1>DID Admin Panel</h1>
+      {status === 'loading' && <p>Loading...</p>}
+      {status === 'unauthenticated' && <LoginButton />}
+      {status === 'authenticated' && (
+        <>
+          <p>Signed in as {session.user?.email}</p>
+        </>
+      )}
```
Comment (victor-heliomar) [inline pages/index.tsx:16]: If here you're validating if the user is unauthenticated, why is needed this logic?

https://github.com/Azordev/did-admin-panel/blob/b0660e259cbc9c9b4afa8bc028255306de16619e/components/LoginButton.tsx#L5-L19

Or it's for another case of use?

File: services/GraphQL/queries/users.ts (hunk 10 above)
Comment (Israel-Laguan) [reply services/GraphQL/queries/users.ts:23]: You right, removed condition on Login

File: pages/index.tsx (hunk 12 above)
Comment (Israel-Laguan) [reply pages/index.tsx:16]: Ah, the' session' use case is quite different from `status`. Please notice `status==='loading'` is not playing well with `if(session)`.

So the idea is: if you need to know when auth is loading, use status, but if loading is not a concern use `session`.

File: views/Login/index.tsx (hunk 8 above)
Comment (Israel-Laguan) [reply views/Login/index.tsx:30]: Ah, I left the case if something is done wrong in DB. I think is like a safety net. Will change the type to "OTHER" and add a note about uniqueness check-in DB for codeMember.

File: views/Login/index.tsx (hunk 9 above)
Comment (Israel-Laguan) [reply views/Login/index.tsx:36]: I erased the condition but left the else in this condition as a safety net if in the future the DB changes.

Comment (Israel-Laguan) [reply views/Login/index.tsx:36]: I think is better to manage the active logic here

File: components/LoginButton.tsx (hunk 11 above)
Comment (Israel-Laguan) [reply components/LoginButton.tsx]: Improved as you suggested

**Comment by e-azocar** [comment]
After clicking Magic Link I received this error
image

**Review by e-azocar — APPROVED** [review]
It's ok the code, only that error
